raster/src/main/scala/geotrellis/raster/rasterize/polygon/FractionalRasterizer.scala

```diff
@@ -0,0 +1,185 @@
+    val xmaxint =  ceil(xmax).toInt
+    val ymaxint =  ceil(ymax).toInt
+
+    // Map coordinates
+    val xminmap = re.gridColToMap(xminint) - re.cellwidth/2
+    val yminmap = re.gridRowToMap(ymaxint) + re.cellheight/2
+    val xmaxmap = re.gridColToMap(xmaxint) - re.cellwidth/2
+    val ymaxmap = re.gridRowToMap(yminint) + re.cellheight/2
+
+    // Envelope around the edge (in map space)
+    val envelope = Polygon(
+      Point(xminmap, yminmap),
+      Point(xminmap, ymaxmap),
+      Point(xmaxmap, ymaxmap),
+      Point(xmaxmap, yminmap),
+      Point(xminmap, yminmap)
+    ).jtsGeom
+
+    // Intersection of envelope and polygon (in map space)
+    val localPoly = poly.jtsGeom.intersection(envelope)
+
+    if (abs(m) <= 1) { // Mostly horizontal
+      var x = xminint; while (x <= xmaxint) {
+        val _y = floor(m * (x + 0.5 - x0) + y0).toInt
+        var i = -1; while (i <= 1) {
+          val y = _y + i
+          val pair = (x, y)
+          val xmap0 = re.gridColToMap(x+0) - re.cellwidth/2
+          val xmap1 = re.gridColToMap(x+1) - re.cellwidth/2
+          val ymap0 = re.gridRowToMap(y+0) + re.cellheight/2
+          val ymap1 = re.gridRowToMap(y+1) + re.cellheight/2
+          val pixel = Polygon(
+            Point(xmap0, ymap0),
+            Point(xmap0, ymap1),
+            Point(xmap1, ymap1),
+            Point(xmap1, ymap0),
+            Point(xmap0, ymap0)
+          ).jtsGeom
+          val fraction = (pixel.intersection(localPoly)).getArea / pixel.getArea
+
+          if (fraction > 0.0) {
+            synchronized {
+              if (!set.contains(pair)) {
+                fn(x, y, fraction)
+                set += ((x, y))
+              }
+            }
+          }
+          i += 1
+        }
+        x += 1
+      }
+    } else { // Mostly vertical
+      val m = (x1 - x0) / (y1 - y0)
+      var y = yminint; while (y <= ymaxint) {
+        val _x = floor(m * (y + 0.5 - y0) + x0).toInt
+        var i = -1; while (i <= 1) {
+          val x = _x + i
+          val pair = (x, y)
+          val xmap0 = re.gridColToMap(x+0) - re.cellwidth/2
+          val xmap1 = re.gridColToMap(x+1) - re.cellwidth/2
+          val ymap0 = re.gridRowToMap(y+0) + re.cellheight/2
+          val ymap1 = re.gridRowToMap(y+1) + re.cellheight/2
+          val pixel = Polygon(
+            Point(xmap0, ymap0),
+            Point(xmap0, ymap1),
+            Point(xmap1, ymap1),
+            Point(xmap1, ymap0),
+            Point(xmap0, ymap0)
+          ).jtsGeom
+          val fraction = (pixel.intersection(localPoly)).getArea / pixel.getArea
+
+          if (fraction > 0.0) {
+            synchronized {
+              if (!set.contains(pair)) {
+                fn(x, y, fraction)
+                set += ((x, y))
+              }
+            }
+          }
+          i += 1
+        }
+        y += 1
+      }
+    }
+  }
+
+  def foreachCellByPolygon(
+    poly: Polygon,
+    re: RasterExtent
+  )(fn: FractionCallback): Unit = {
```

Comment:
We should use a trait instead of Function3, because Function3 isn't specialized. The way we get around this for tile methods is via macros. We would eventually want to macroize this as well, but that can be a TODO.